trusted/pal/src/pal.rs

```diff
@@ -0,0 +1,174 @@
+    ///
+    /// # Returns
+    ///
+    /// An error if the signature cannot be verifed or is invalid, a success otherwise.
+    fn verify(&self, data: &[u8], signature: &[u8]) -> Result<()>;
+
+    /// Gets serialized public key used for signing by the trusted application.
+    /// The signing key is generated at the start of the trusted application and
+    /// remains unchanged through the lifetime.
+    ///
+    /// # Note
+    ///
+    /// Public signing key can be used to derive a trusted application identity.
+    /// For example, a trusted application that represents a node in Raft cluster
+    /// running inside of a group trusted hosts, it is important that Raft node
+    /// identity cannot be forged, does not require coordination to pick one and
+    /// has low collision chance. Hash of the public signing key is an identity
+    /// mechanism that is compliant with these requirements.
+    fn public_signing_key(&self) -> Vec<u8>;
+}
+
+/// Represents a trusted host, abstracting away the details of how trusted
+/// application is hosted (e.g. trusted host can be restricted kernel bare
+/// metal based encrypted virtual machine or a linux kernel docker based
+/// encrypted virtual machine).
+pub trait Host {
+    /// Gets attestation that represens a set of claims about the trusted application
+    /// and its host. The attestation is generated after the trusted application and
+    /// host are fully initialiezed, the attestation doesn't change throughout the
+    /// lifetime.
+    fn get_self_attestation(&self) -> Box<dyn Attestation>;
+
+    /// Gets serialized configuration that stays immutable through the lifetime of
+    /// the trusted application.
+    fn get_self_config(&self) -> Vec<u8>;
+
+    /// Sends messages through the communication channel that connects the trusted
+    /// application to the untrusted launcher.
+    ///
+    /// # Argumnets
+    ///
+    /// * `messages` - A set of messsages to send through the channel.
+    ///
+    /// # Returns
+    ///
+    /// Error if the communication channel is irrepairably broken, a success otherise.
+    fn send_messages(&mut self, messages: &[Message]) -> Result<()>;
+
+    /// Attempts to deserialize peer attestation and perform initial verification
+    /// of the attestation. The application specific verification (e.g. ensuring
+    /// that attestation points to a particular binary and configuration) is the
+    /// responsibility of the trusted application.
+    ///
+    /// # Arguments
+    ///
+    /// * `peer_attestation` - A serialized attestation of a peer trusted application.
+    ///
+    /// # Returns
+    ///
+    /// A deserialized and initially verified peer attestation if success, an error
+    /// otherwise. An error may be caused by malformed serialized representation or
+    /// by failing verification.
+    fn verify_peer_attestation(&self, peer_attestation: &[u8]) -> Result<Box<dyn Attestation>>;
```

Comment:
The result of this is only constrained by the Attestation trait, so I'm not sure how it's expected to be used, could you elaborate?

Reply:
Let's imagine there is a single node Raft cluster. We are trying to add another node to the cluster by sending configuration change to the leader. The leader tries to send append entries to the new node. Before doing so, we need to establish trust and secure channel. The leader obtains its attestation from the host (platform), generates its public key for the HPKE and signs it, sends both in an initiate handshake message. The follower receives the initiate handshake message, gets serialized leader attestation and asks the host (platform) to verify (check that the DICE chain is valid) and deserialize it. Now that the follower has the deserialized attestation containing public signing key it can verify signed public key of the leader for the HPKE. Afterwards the follower performs symmetric action (gets its attestation, generates HPKE response and sign it, serializes attestation and sends it and signed HPKE response to the leader). Hope this helps.